Enable passing pycapnp builders/readers to C++ code

To integrate the Python and C++ serialization code, we need to be able to pass a builder or reader for some part of a Cap'n Proto struct from the Python code to the C++ code.

Part of #1449

hi,
I'm a newbie and I have problem when I install pycapnp. I download pycapnp-0.5.1.tar.gz and the instrument says I need to install capnproto-c++-0.5.0 first, but when I type the commad 'sudo make install', I encounter the problems as follows:
depbase=`echo src/capnp/compiler/module-loader.o | sed 's|[^/]*$|.deps/&|;s|\.o$||'`;\
	g++ -std=gnu++11 -DHAVE_CONFIG_H -I.    -I./src -I./src -DKJ_HEADER_WARNINGS -DCAPNP_HEADER_WARNINGS -DCAPNP_INCLUDE_DIR='"/usr/local/include"' -pthread -O2 -DNDEBUG -pthread -MT src/capnp/compiler/module-loader.o -MD -MP -MF $depbase.Tpo -c -o src/capnp/compiler/module-loader.o src/capnp/compiler/module-loader.c++ &&\
	mv -f $depbase.Tpo $depbase.Po
In file included from ./src/capnp/layout.h:37:0,
                 from ./src/capnp/generated-header-support.h:31,
                 from ./src/capnp/compiler/grammar.capnp.h:7,
                 from src/capnp/compiler/compiler.h:29,
                 from src/capnp/compiler/module-loader.h:29,
                 from src/capnp/compiler/module-loader.c++:22:
./src/kj/memory.h:213:28: error: expected ‘;’ at end of member declaration
./src/kj/memory.h:213:32: error: expected unqualified-id before ‘{’ token
./src/kj/memory.h:213:48: error: expected ‘;’ at end of member declaration
./src/kj/memory.h:214:36: error: expected ‘;’ at end of member declaration
./src/kj/memory.h:214:44: error: expected unqualified-id before ‘{’ token
./src/kj/memory.h:214:60: error: expected ‘;’ at end of member declaration
./src/kj/memory.h:215:29: error: expected ‘;’ at end of member declaration
./src/kj/memory.h:215:19: error: ‘kj::Own<T>&& kj::_::OwnOwn<T>::operator*()’ cannot be overloaded
./src/kj/memory.h:213:18: error: with ‘kj::Own<T>& kj::_::OwnOwn<T>::operator*()’
./src/kj/memory.h:215:34: error: expected unqualified-id before ‘{’ token
./src/kj/memory.h:215:58: error: expected ‘;’ at end of member declaration
./src/kj/memory.h:216:37: error: expected ‘;’ at end of member declaration
./src/kj/memory.h:216:25: error: ‘const kj::Own<T>&& kj::_::OwnOwn<T>::operator*() const’ cannot be overloaded
./src/kj/memory.h:214:24: error: with ‘const kj::Own<T>& kj::_::OwnOwn<T>::operator*() const’
./src/kj/memory.h:216:46: error: expected unqualified-id before ‘{’ token
./src/kj/memory.h:216:70: error: expected ‘;’ at end of member declaration
In file included from ./src/capnp/compiler/grammar.capnp.h:7:0,
                 from src/capnp/compiler/compiler.h:29,
                 from src/capnp/compiler/module-loader.h:29,
                 from src/capnp/compiler/module-loader.c++:22:
./src/capnp/generated-header-support.h: In constructor ‘constexpr capnp::_::RawBrandedSchema::Binding::Binding(uint8_t, uint16_t, const capnp::_::RawBrandedSchema*)’:
./src/capnp/generated-header-support.h:82:27: error: uninitialized member ‘capnp::_::RawBrandedSchema::Binding::<anonymous>’ in ‘constexpr’ constructor
./src/capnp/generated-header-support.h: In constructor ‘constexpr capnp::_::RawBrandedSchema::Binding::Binding(uint8_t, uint16_t, uint64_t, uint16_t)’:
./src/capnp/generated-header-support.h:86:29: error: uninitialized member ‘capnp::_::RawBrandedSchema::Binding::<anonymous>’ in ‘constexpr’ constructor
./src/capnp/generated-header-support.h: In constructor ‘constexpr capnp::_::RawBrandedSchema::Binding::Binding(uint8_t, uint16_t, uint16_t)’:
./src/capnp/generated-header-support.h:89:55: error: uninitialized member ‘capnp::_::RawBrandedSchema::Binding::<anonymous>’ in ‘constexpr’ constructor
make: *** [src/capnp/compiler/module-loader.o] Error 1

Can you tell me what shoud I do to sovle it ? Hoping for your reply. Thank you so much.

--- a/.gitignore
+++ b/.gitignore
@@ -66,8 +66,9 @@ extensions/core/
 experimental/
 
 # nupic bindings swig libraries
-nupic/bindings/
-~nupic/bindings/*.i
+nupic/bindings/*.cxx
+nupic/bindings/*.py
+nupic/bindings/*.so
 
 # Pip stuff
 dist/

--- a/extensions/py_support/PyCapnp.hpp
+++ b/extensions/py_support/PyCapnp.hpp
@@ -1,0 +1,100 @@
+/* ---------------------------------------------------------------------
+ * Numenta Platform for Intelligent Computing (NuPIC)
+ * Copyright (C) 2015, Numenta, Inc.  Unless you have an agreement
+ * with Numenta, Inc., for a separate license for this software code, the
+ * following terms and conditions apply:
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see http://www.gnu.org/licenses.
+ *
+ * http://numenta.org/licenses/
+ * ---------------------------------------------------------------------
+ */
+
+// This file contains utility functions for converting from pycapnp schema to
+// compiled in schema. It requires linking to both libcapnp and libcapnpc.
+
+#ifndef NTA_PY_CAPNP_HPP
+#define NTA_PY_CAPNP_HPP
+
+#include <Python.h>
+
+#include <capnp/any.h>
+#include <capnp/dynamic.h>
+#include <capnp/message.h>
+#include <capnp/schema-parser.h>
+
+namespace nupic
+{
+
+  struct pycapnp_SchemaParser {
+    PyObject_HEAD
+    void *__pyx_vtab;
+     ::capnp::SchemaParser *thisptr;
+    PyObject *modules_by_id;
+  };
+
+  struct pycapnp_DynamicStructBuilder {
+    PyObject_HEAD
+    void *__pyx_vtab;
+     ::capnp::DynamicStruct::Builder thisptr;
+    PyObject *_parent;
+    int is_root;
+    int _is_written;
+    PyObject *_schema;
+  };
+
+  struct pycapnp_DynamicStructReader {
+    PyObject_HEAD
+    void *__pyx_vtab;
+     ::capnp::DynamicStruct::Reader thisptr;
+    PyObject *_parent;
+    int is_root;
+    PyObject *_obj_to_pin;
+    PyObject *_schema;
+  };
+
+  template<class T>
+  typename T::Builder getBuilder(PyObject* pyBuilder)
+  {
+    PyObject* capnpModule = PyImport_AddModule("capnp.lib.capnp");
+    PyObject* pySchemaParser = PyObject_GetAttrString(capnpModule,
+                                                      "_global_schema_parser");
+    pycapnp_SchemaParser* schemaParser = (pycapnp_SchemaParser*)pySchemaParser;
+    schemaParser->thisptr->loadCompiledTypeAndDependencies<T>();
+
+    pycapnp_DynamicStructBuilder* dynamicStruct =
+        (pycapnp_DynamicStructBuilder*)pyBuilder;
+    capnp::DynamicStruct::Builder& builder = dynamicStruct->thisptr;
+    typename T::Builder proto = builder.as<T>();
+    return proto;
+  }
+
+  template<class T>
+  typename T::Reader getReader(PyObject* pyReader)
+  {
+    PyObject* capnpModule = PyImport_AddModule("capnp.lib.capnp");
+    PyObject* pySchemaParser = PyObject_GetAttrString(capnpModule,
+                                                      "_global_schema_parser");
+    pycapnp_SchemaParser* schemaParser = (pycapnp_SchemaParser*)pySchemaParser;
+    schemaParser->thisptr->loadCompiledTypeAndDependencies<T>();
+
+    pycapnp_DynamicStructReader* dynamicStruct =
+        (pycapnp_DynamicStructReader*)pyReader;
+    capnp::DynamicStruct::Reader& reader = dynamicStruct->thisptr;
+    typename T::Reader proto = reader.as<T>();
+    return proto;
+  }
+
+}  // namespace nupic
+
+#endif  // NTA_PY_CAPNP_HPP

--- a/nupic/bindings/algorithms.i
+++ b/nupic/bindings/algorithms.i
@@ -1,6 +1,6 @@
 /* ---------------------------------------------------------------------
  * Numenta Platform for Intelligent Computing (NuPIC)
- * Copyright (C) 2013, Numenta, Inc.  Unless you have an agreement
+ * Copyright (C) 2013-2015, Numenta, Inc.  Unless you have an agreement
  * with Numenta, Inc., for a separate license for this software code, the
  * following terms and conditions apply:
  *
@@ -99,18 +99,20 @@ _ALGORITHMS = _algorithms
 #include <nupic/algorithms/SpatialPooler.hpp>
 #include <nupic/algorithms/FlatSpatialPooler.hpp>
 
+#include <nupic/algorithms/Cell.hpp>
 #include <nupic/algorithms/Cells4.hpp>
 #include <nupic/algorithms/ClassifierResult.hpp>
-#include <nupic/algorithms/FastClaClassifier.hpp>
-#include <nupic/algorithms/SegmentUpdate.hpp>
-#include <nupic/algorithms/OutSynapse.hpp>
-#include <nupic/algorithms/InSynapse.hpp>
-#include <nupic/algorithms/Cell.hpp>
-
 #include <nupic/algorithms/Connections.hpp>
+#include <nupic/algorithms/FastClaClassifier.hpp>
+#include <nupic/algorithms/InSynapse.hpp>
+#include <nupic/algorithms/OutSynapse.hpp>
+#include <nupic/algorithms/SegmentUpdate.hpp>
+
+#include <nupic/proto/SpatialPoolerProto.capnp.h>
 
 #include <numpy/arrayobject.h>
 #include <py_support/NumpyVector.hpp>
+#include <py_support/PyCapnp.hpp>
 #include <py_support/PythonStream.hpp>
 #include <py_support/PyHelpers.hpp>
 
@@ -1900,6 +1902,19 @@ inline PyObject* generate2DGaussianSample(nupic::UInt32 nrows, nupic::UInt32 nco
   {
     PyArrayObject* x = (PyArrayObject*) py_x;
     self->stripUnlearnedColumns((nupic::UInt*) x->data);
+  }
+
+  inline void write(PyObject* pyBuilder) const
+  {
+    SpatialPoolerProto::Builder proto =
+        getBuilder<SpatialPoolerProto>(pyBuilder);
+    self->write(proto);
+  }
+
+  inline void read(PyObject* pyReader)
+  {
+    SpatialPoolerProto::Reader proto = getReader<SpatialPoolerProto>(pyReader);
+    self->read(proto);
   }
 
   void loadFromString(const std::string& inString)

--- a/setup.py
+++ b/setup.py
@@ -341,7 +341,11 @@ def getExtensionModules(nupicCoreReleaseDir, platform, bitness):
   commonLinkFlags = [
     "-m" + bitness,
     "-fPIC",
-    "-L" + pythonPrefix + "/lib"]
+    "-L" + pythonPrefix + "/lib",
+    "-lkj",
+    "-lcapnp",
+    "-lcapnpc",
+  ]
 
   commonLibraries = [
     "dl",

--- a/setup.py
+++ b/setup.py
@@ -342,6 +342,7 @@ def getExtensionModules(nupicCoreReleaseDir, platform, bitness):
     "-m" + bitness,
     "-fPIC",
     "-L" + pythonPrefix + "/lib",
+    "-L" + nupicCoreReleaseDir + "/lib",
     "-lkj",
     "-lcapnp",
     "-lcapnpc",

